src/main/java/uk/gov/hmcts/ccd/service/MessagePublisherRunnable.java

```diff
@@ -0,0 +1,82 @@
+import uk.gov.hmcts.ccd.data.MessageQueueCandidateEntity;
+import uk.gov.hmcts.ccd.data.MessageQueueCandidateRepository;
+
+import java.time.LocalDateTime;
+import java.util.List;
+
+import static com.google.common.collect.Lists.newArrayList;
+
+@Slf4j
+public class MessagePublisherRunnable implements Runnable {
+
+    private MessageQueueCandidateRepository messageQueueCandidateRepository;
+    private JmsTemplate jmsTemplate;
+    private PublishMessageTask publishMessageTask;
+    private String logPrefix;
+
+    public MessagePublisherRunnable(MessageQueueCandidateRepository messageQueueCandidateRepository,
+                                    JmsTemplate jmsTemplate,
+                                    PublishMessageTask publishMessageTask) {
+        this.messageQueueCandidateRepository = messageQueueCandidateRepository;
+        this.jmsTemplate = jmsTemplate;
+        this.publishMessageTask = publishMessageTask;
+        this.logPrefix = String.format("[Message Type - %s]", publishMessageTask.getMessageType());
+    }
+
+    @Override
+    public void run() {
+        log.debug(String.format("%s Starting publish message task", logPrefix));
+        processUnpublishedMessages(PageRequest.of(0, publishMessageTask.getBatchSize()), newArrayList());
+        deletePublishedMessages();
+        log.debug(String.format("%s Completed publish message task", logPrefix));
+    }
+
+    private void processUnpublishedMessages(Pageable pageable,
+                                            List<MessageQueueCandidateEntity> processedEntities) {
+        Slice<MessageQueueCandidateEntity> unpublishedMessagesPaginated = null;
+        boolean hasError = false;
+
+        try {
+            unpublishedMessagesPaginated = messageQueueCandidateRepository
+                .findUnpublishedMessages(publishMessageTask.getMessageType(), pageable);
+            publishMessages(unpublishedMessagesPaginated, processedEntities);
+        } catch (Exception e) {
+            log.error(String.format("%s Error encountered during processing of "
+                + "unpublished messages", logPrefix), e);
+            hasError = true;
+        }
+
+        if (!hasError && unpublishedMessagesPaginated.hasNext()) {
+            processUnpublishedMessages(unpublishedMessagesPaginated.nextPageable(), processedEntities);
+        } else if (!processedEntities.isEmpty()) {
+            messageQueueCandidateRepository.saveAll(processedEntities);
+            log.info(String.format("%s Published %s messages to destination '%s'",
+                logPrefix, processedEntities.size(), publishMessageTask.getDestination()));
+        }
+    }
+
+    private void publishMessages(Slice<MessageQueueCandidateEntity> messagesToPublish,
+                                 List<MessageQueueCandidateEntity> processedEntities) {
+        messagesToPublish.get().forEach(entity -> {
+            jmsTemplate.convertAndSend(publishMessageTask.getDestination(), entity.getMessageInformation());
```

Comment:
if message publishing is failed, then we need to stop processing other messages and have to re-try the failed message until that succeeds.. seems like this logic is missing

Reply:
And, messages need to be processed in order, is this method guarantees the order?